[#6] [UI] As a user, I can see the Sign In page

manh-t opened the pull request:
Close #6 

## What happened 👀

- Ignore `tailwind` in `stylelint`
- Remove sample code
- Create SignInScreen
- Write unittest.

## Insight 📝

I use `tailwind` to apply the styles to the components, and I create some reusable attributes to work with `tailwind`.

## Proof Of Work 📹

![image](https://github.com/manh-t/react-survey/assets/60863885/9afcc33b-03e6-4ce8-b2ed-49740980b96d)

Check the Netlify link below.

manh-t committed:
[#6] Create SignInScreen

manh-t commented:
<!-- NETLIFY DEPLOY COMMENT GENERATED BY ACTIONS_NETLIFY - APP ID SHA256: 8b679132dbd60f1ef5a25fa0d2b15d442565e859c91e04b6f53bfc055c0407ff -->
🚀 Deployed on https://649e50e8f120a02951095552--manh-react-survey.netlify.app

manh-t committed:
[#6] Add Forgot text to the password field

manh-t committed:
[#6] Ignore tailwind stylelint

manh-t committed:
[#6] Create custom ElevatedButton

manh-t committed:
[#6] Add unittest for SignInScreen

manh-t committed:
[#6] Fix UI test

manh-t commented:
<!--
  0 failure: 
  0 warning: 
  1 messages
  1 markdown notices
  DangerID: danger-id-Danger;
-->



<table>
  <thead>
    <tr>
      <th width="50"></th>
      <th width="100%" data-danger-table="true">Messages</th>
    </tr>
  </thead>
  <tbody><tr>
      <td>:book:</td>
      <td>Congrats, coverage is good</td>
    </tr>
  </tbody>
</table>

## Coverage in All Files
File | Line Coverage | Statement Coverage | Function Coverage | Branch Coverage
---- | ------------: | -----------------: | ----------------: | --------------:
[src/App.tsx](../blob/e277b2e/src/App.tsx) | (3/3) 100% | (3/3) 100% | (1/1) 100% | (0/0) NaN%
[src/index.tsx](../blob/e277b2e/src/index.tsx) | (3/3) 100% | (3/3) 100% | (0/0) NaN% | (0/0) NaN%
[src/../index.tsx](../blob/e277b2e/src/components/ElevatedButton/index.tsx) | (3/3) 100% | (3/3) 100% | (1/1) 100% | (0/0) NaN%
[src/../TextInput/index.tsx](../blob/e277b2e/src/components/TextInput/index.tsx) | (3/3) 100% | (3/3) 100% | (1/1) 100% | (2/2) 100%
[src/lib/requestManager.ts](../blob/e277b2e/src/lib/requestManager.ts) | (5/5) 100% | (5/5) 100% | (2/2) 100% | (1/1) 100%
[src/routes/index.tsx](../blob/e277b2e/src/routes/index.tsx) | (1/1) 100% | (1/1) 100% | (0/0) NaN% | (0/0) NaN%
[src/screens/SignIn/index.tsx](../blob/e277b2e/src/screens/SignIn/index.tsx) | (3/3) 100% | (3/3) 100% | (1/1) 100% | (0/0) NaN%
Total | (21/21) 100% | (21/21) 100% | (6/6) 100% | (3/3) 100%

<p align="right">
  Generated by :no_entry_sign: <a href="https://danger.systems/js">dangerJS</a> against c5133e8423bb3bf6010e64ea3847e2b36a1878f7
</p>

nvminhtue commented on .eslintrc.js:
```diff
@@ -1,5 +1,10 @@
+    '@nimblehq/eslint-config-nimble-react',
+    '@nimblehq/eslint-config-nimble-testing',
+    'plugin:cypress/recommended',
+    'prettier',
```
we can remove it as already added in `@nimblehq/eslint-config-nimble-react`, we don't need to override

nvminhtue commented on src/components/ElevatedButton/index.tsx:
```diff
@@ -0,0 +1,23 @@
+    <button
+      type="button"
+      className={
+        'bg-white text-black-chinese font-bold text-regular rounded-[10px] focus:outline-none focus:shadow-outline h-14' +
+        (isFullSize ? ' w-full' : '')
```
let's try to concat by using `` string with dynamic variable

nvminhtue commented on src/screens/SignIn/index.tsx:
```diff
@@ -0,0 +1,69 @@
+  return (
+    <div
+      className="sign-in bg-cover min-h-screen flex flex-col justify-center items-center"
+      style={{ backgroundImage: 'url(' + background + ')' }}
```
```suggestion
      style={{ backgroundImage: `url '${background}'` }}
```

nvminhtue replied:
But can we try [this](https://tailwindcss.com/docs/background-image) out?

manh-t replied:
@nvminhtue ah thanks, I will update it 👏

nvminhtue commented on src/screens/SignIn/index.tsx:
```diff
@@ -0,0 +1,69 @@
+              id: 'form-sign-in-email',
+              required: true,
+              type: 'email',
+              'data-test-id': signInScreenTestIds.emailField,
```
```suggestion
              data-test-id: signInScreenTestIds.emailField,
```

manh-t replied:
@nvminhtue unfortunately, I cannot remove the single quotes due to the error 🙏 
![image](https://github.com/manh-t/react-survey/assets/60863885/57605370-8ee6-4b13-9737-78498f290463)

nvminhtue commented:
![image](https://github.com/manh-t/react-survey/assets/42540264/42067071-59ef-42b5-889e-cab6fea3c1ee)
Could you help me to clarify it?

manh-t committed:
[#6] Apply background opacity when focus text field and fix autofill issue

manh-t commented:
> ![image](https://user-images.githubusercontent.com/42540264/246813473-42067071-59ef-42b5-889e-cab6fea3c1ee.png) Could you help me to clarify it?

Thanks for catching the issue. This is the issue related to autofill. Fixed in [5c478f2](https://github.com/manh-t/react-survey/pull/42/commits/5c478f27a1d3505c73d5809f447bc9c1ba77a217) 🙏 
![image](https://github.com/manh-t/react-survey/assets/60863885/89169a3a-a227-4587-bc4f-a354c1c6eaff)

nvminhtue commented:
The spacing seems closer compared to the design. Please take a look at that and appropriately adjust.
The font also different, when the one in the design is (`Neuzeit S LT Std`)
![image](https://github.com/manh-t/react-survey/assets/42540264/a1c9ac15-52a3-41e4-bdca-008f65c9bfde)

manh-t committed:
[#6] Add the letter spacing to the components

manh-t commented:
> The spacing seems closer compared to the design. Please take a look at that and appropriately adjust. The font also different, when the one in the design is (`Neuzeit S LT Std`) ![image](https://user-images.githubusercontent.com/42540264/247141597-a1c9ac15-52a3-41e4-bdca-008f65c9bfde.png)

I adjusted the spacing ✅ 

Regarding the font, the default font is currently `Neuzeit S LT Std`
![image](https://github.com/manh-t/react-survey/assets/60863885/3d8432a9-94b7-48b4-bc04-fbc2395e5e03)

But the letter spacing is different, so I added the letter spacing in 6be17d4
Left (design) vs Right (actual)
![image](https://github.com/manh-t/react-survey/assets/60863885/3a60ba6e-debc-4bb0-a513-06d2b6dbd3e6)

nvminhtue reviewed (approved):
👏

tyrro commented on src/screens/SignIn/index.tsx:
```diff
@@ -0,0 +1,65 @@
+            Forgot?
+          </button>
+        </div>
+        <ElevatedButton isFullSize={true} type="submit" data-test-id={signInScreenTestIds.signInButton}>
```
attributes without values are interpreted as a boolean true, so we can write:
 
```suggestion
        <ElevatedButton isFullSize type="submit" data-test-id={signInScreenTestIds.signInButton}>
```

manh-t replied:
@tyrro resolved in [1aa1c5e](https://github.com/manh-t/react-survey/pull/42/commits/1aa1c5e5c8d8d2c9cda3ac696056fd9e24c867d1)

tyrro commented on src/assets/stylesheets/application.scss:
```diff
@@ -1,3 +1,29 @@
-@import 'tailwindcss/base';
-@import 'tailwindcss/components';
-@import 'tailwindcss/utilities';
+@tailwind base;
+@tailwind components;
+@tailwind utilities;
```
could we move the tailwind-specific lines to a separate file `base/tailwind.scss` and import it in `base`? The idea is to keep `application.scss` clear and concise.

manh-t replied:
@tyrro resolved in [1aa1c5e](https://github.com/manh-t/react-survey/pull/42/commits/1aa1c5e5c8d8d2c9cda3ac696056fd9e24c867d1)

tyrro commented on src/components/ElevatedButton/index.tsx:
could we have unit tests for the components? Or do you plan to do it separately?

manh-t replied:
@tyrro ah let me add the unittest for components 🙏

manh-t replied:
@tyrro added in [1aa1c5e](https://github.com/manh-t/react-survey/pull/42/commits/1aa1c5e5c8d8d2c9cda3ac696056fd9e24c867d1), thanks a lot 😄

manh-t committed:
[#6] Add unittest for the widget components

tyrro commented on src/assets/stylesheets/application.scss:
```diff
@@ -1,3 +1,25 @@
+  transition: background-color 5000s ease-in-out 0s;
+
+  background-color: #fff;
+}
```
could we move these to `tailwind.scss` as these are specific to tailwind?

manh-t replied:
@tyrro resolved in [a31c486](https://github.com/manh-t/react-survey/pull/42/commits/a31c486e6aabdb861be574b7d76fd7ec5686a78d)

tyrro commented on src/assets/stylesheets/base/_tailwindcss.scss:
maybe we could just name this file to simply `_tailwind.scss`?

tyrro commented on src/components/ElevatedButton/index.test.tsx:
```diff
@@ -0,0 +1,39 @@
+
+describe('ElevatedButton', () => {
+  const label = 'Sign in';
+  const dataTestId = 'sign-in_button';
```
let's have an empty line before the start of  the `it`  block!

manh-t replied:
@tyrro resolved in [a31c486](https://github.com/manh-t/react-survey/pull/42/commits/a31c486e6aabdb861be574b7d76fd7ec5686a78d)

tyrro commented on src/components/ElevatedButton/index.tsx:
```diff
@@ -0,0 +1,21 @@
+  return (
+    <button
+      type="button"
+      className={`bg-white text-black-chinese font-bold text-regular tracking-survey-tight rounded-[10px] focus:outline-none focus:shadow-outline h-14 ${fullWidth}`}
```
let's use [classnames](https://github.com/JedWatson/classnames) library here, it'll simplify using classnames a lot 😉 
For example, this code can be re-written as:
```suggestion
            className={classNames(
              'bg-white text-black-chinese font-bold text-regular tracking-survey-tight rounded-[10px] focus:outline-none focus:shadow-outline h-14',
              { 'w-full': isFullSize },
            )}
```

tyrro replied:
Also, we could extract the class names to a variable for better readability:

```js
const DEFAULT_CLASS_NAMES = 'bg-white text-black-chinese font-bold text-regular tracking-survey-tight rounded-[10px] focus:outline-none focus:shadow-outline h-14'
```

manh-t replied:
TIL, thanks, @tyrro

tyrro commented on src/components/TextInput/index.test.tsx:
```diff
@@ -0,0 +1,52 @@
+  const inputDataTestId = 'text-input';
+  const inputType = 'email';
+  const inputPlaceholder = 'input placeholder';
+  it('renders a text input with a label', () => {
```
```suggestion
 
   it('renders a text input with a label', () => {
```

tyrro commented on src/components/ElevatedButton/index.test.tsx:
```diff
@@ -0,0 +1,39 @@
+    expect(button).toHaveClass('w-full');
+  });
+
+  it('renders a wrap-content button', () => {
```
we could simply say that

```suggestion
  it('does NOT render a full-width button', () => {
```

It makes it easier to read, what do you think?

tyrro commented on src/components/ElevatedButton/index.tsx:
```diff
@@ -0,0 +1,21 @@
+
+interface ElevatedButtonProps extends React.ButtonHTMLAttributes<HTMLButtonElement> {
+  children: React.ReactNode;
+  isFullSize: boolean;
```
I am whether it'll be more appropriate if we name this prop `isFullWidth`?

manh-t replied:
@tyrro make sense, let me update it 🙏

manh-t replied:
@tyrro resolved in [a31c486](https://github.com/manh-t/react-survey/pull/42/commits/a31c486e6aabdb861be574b7d76fd7ec5686a78d)

tyrro commented on src/components/TextInput/index.test.tsx:
```diff
@@ -0,0 +1,52 @@
+    };
+    render(<TextInput {...inputProps} />);
+
+    expect(screen.queryByRole('label')).not.toBeInTheDocument();
```
[toBeInTheDocument](https://github.com/testing-library/jest-dom#tobeinthedocument) passes if the element is `display: none` or `visibility: hidden`, which we don't want in this case, so could we use `toBeVisible`?

```suggestion
    expect(screen.queryByRole('label')).not.toBeVisible();
```

manh-t replied:
@tyrro actually when the label is empty, the `<label>` element will not be generated so `toBeInTheDocument()` is my expected result in this case 😄 

I also try to use `toBeVisible()` but it doesn't work with the does-not-exist element 🙏 
![image](https://github.com/manh-t/react-survey/assets/60863885/90ae4a62-9d0d-458a-aa8e-6465cf965aeb)

tyrro replied:
ah I see, we are testing a non-existing element!

tyrro commented on src/screens/SignIn/index.tsx:
```diff
@@ -0,0 +1,65 @@
+                type: 'password',
+                'data-test-id': signInScreenTestIds.passwordField,
+              }}
+              extraClassName="pr-16"
```
we could just name it as `className`,  it's a common naming convention, for example in React codebase:

https://github.com/facebook/react/blob/a1c62b8a7635c0bc51e477ba5437df9be5a9e64f/packages/react-devtools-shared/src/devtools/views/Icon.js#L32

manh-t replied:
got it @tyrro 🚀

manh-t replied:
@tyrro resolved in [a31c486](https://github.com/manh-t/react-survey/pull/42/commits/a31c486e6aabdb861be574b7d76fd7ec5686a78d)

tyrro commented on tailwind.config.js:
```diff
@@ -2,7 +2,29 @@
+        'sign-in': "url('assets/images/illustrations/background-sign-in.png')",
+      },
+      letterSpacing: {
+        'survey-tighest': '-1px',
```
hopefully, we don't have to invent any more `tight`-adjectives lol 😆 
btw  the naming [goes with the tailwind world](https://tailwindcss.com/docs/letter-spacing) as well 👍🏿

manh-t replied:
@tyrro I checked the Figma and added all customized letter spacings here, so that's all 😄

manh-t committed:
[#6] Apply classnames package for util and small refactors

tyrro commented on src/components/ElevatedButton/index.tsx:
```diff
@@ -0,0 +1,20 @@
+const ElevatedButton = ({ children, isFullWidth, ...attributes }: ElevatedButtonProps): JSX.Element => {
+  const DEFAULT_CLASS_NAMES =
+    'bg-white text-black-chinese font-bold text-regular tracking-survey-tight rounded-[10px] focus:outline-none focus:shadow-outline h-14';
+  return (
```
let's have an empty line before the `return` statement 🙌🏿

tyrro commented on src/components/TextInput/index.tsx:
```diff
@@ -0,0 +1,41 @@
+const TextInput = ({ label, labelDataTestId, inputAttributes, classNames }: TextInputProps): JSX.Element => {
+  const DEFAULT_CLASS_NAMES =
+    'block appearance-none bg-white bg-opacity-[.18] rounded-[12px] w-full h-14 focus:outline-none focus:ring-transparent px-3 mt-2 text-white text-regular tracking-survey-tight focus:bg-opacity-30';
+  return (
```
ditto about the empty line!

tyrro commented on src/components/TextInput/index.tsx:
```diff
@@ -0,0 +1,41 @@
+import React from 'react';
+
+import classnames from 'classnames';
```
let's import `classNames`, it's similar to `classname` but goes with the camel case naming of JS! 😄 

```suggestion
import classNames from 'classnames';
```

tyrro commented on src/screens/SignIn/index.tsx:
```diff
@@ -0,0 +1,65 @@
+import React from 'react';
+
+import nimbleLogoWhite from 'assets/images/icons/nimble-logo-white.svg';
```
I think we are going to import a lot of assets in the coming days:
what about we set the path in `tsconfig.json` so that we can import it like this:

`import nimbleLogoWhite from 'icons/nimble-logo-white.svg';`

```json
"paths": {
        "icons/*": ["assets/images/icons*"]
 }
 ```
 
 You can do it in a separate PR if you want as you probably need to update `jest.config.js` as well!

manh-t replied:
I have tried this but getting error when running the application 😞 I will update it in [another ticket](https://github.com/manh-t/react-survey/issues/60) 🙏 
![image](https://github.com/manh-t/react-survey/assets/60863885/b14934d6-3504-45e3-80a6-a866d6050319)

tyrro reviewed (approved):
I've some last comments but the changes now LGTM! Well done!

manh-t committed:
[#6] Small refactor
